### PR TITLE
ENH: Aggregate Expert

### PR DESCRIPTION
**Pull request by waldbauer-certat**
This expert is able to aggregate events with a defined threshold.

How does it work?
The expert will use specified fields from an event as identifier.
Then stores how often the event occur and if the threshold has
been reached within its set timespan, it will push the event to
the queue, else it will just get dropped.

### CODE REVIEW

File: intelmq/tests/bots/experts/aggregate/testdata/events.csv
Comment (wagner-certat) [inline intelmq/tests/bots/experts/aggregate/testdata/events.csv:1]: The file name is CSV, but the content is JSON

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:19]: Either
- It it should be noted in the docs, that the TTL must be higher than the timespan, otherwise data is lost
- or the TTL is not used at all

File: docs/user/bots.rst
Comment (wagner-certat) [inline docs/user/bots.rst:1767]: this section is missing the parameters fields, threshold and timespan

File: intelmq/bots/experts/aggregate/expert.py
Comment (waldbauer-certat) [reply intelmq/bots/experts/aggregate/expert.py:19]: TTL is currently not used at all. Removing it

File: intelmq/tests/bots/experts/aggregate/test_expert.py
Comment (wagner-certat) [inline intelmq/tests/bots/experts/aggregate/test_expert.py:39]:
```suggestion
```

File: docs/user/bots.rst
Comment (wagner-certat) [inline docs/user/bots.rst:1769]: Please add, that the TTL is not used

**Review by schacht-certat — DISMISSED** [review]
Please add license headers to the files.
Please move the changes that are not directly part of the aggregate expert to separate commits or PRs - i.e. the enhancement in the github workflow can easily be moved to a separate PR and merged in no time.
Can you explain the change in the `intelmq/lib/bot.py`?

**Comment by waldbauer-certat** [comment]
This PR uses #1968, dont merge before the other one got merged.

File: intelmq/tests/bots/experts/aggregate/testdata/events.txt
Comment (wagner-certat) [inline intelmq/tests/bots/experts/aggregate/testdata/events.txt:1]: `93.184.216.34` is not a reserved address, please use another one (the rfc1918 expert has a compact list of neutral network ranges)

File: docs/user/bots.rst
Comment (wagner-certat) [inline docs/user/bots.rst:1797]:
```suggestion
If you're running on low traffic ( less events within the given timestamp ) we'd recommend reloading or restarting the bot via cronjob
```
Maybe add an example cron-line?

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:61]:
```suggestion
        cache_id = f"aggregate.{message_hash}"
```

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:63]: Why not this way?
```suggestion
        if self.cache_exists(cache_id):
```

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:66]:
```suggestion
            pipe.hincrby(name=cache_id, key="c", amount=int(event.get('extra.count', 1)))
```
The same scheme can be applied to various other lines

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:60]: The default filter for the hash is blacklist, so you are computing a hash with `self.fields` removed. This looks unintuitive to me, is this intentional?

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:67]: Please add some inline comments what these commands do

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:43]: While comparing time as ISO-formatted strings is possible, is this conversion step necessary?

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [inline intelmq/bots/experts/aggregate/expert.py:24]:
```suggestion
```
Is already in the `CacheMixin`

**Review by wagner-certat — CHANGES_REQUESTED** [review]
only a preliminary review

Please add some more details in the docs explaining what the bot does

File: intelmq/tests/bots/experts/aggregate/testdata/events.txt
Comment (waldbauer-certat) [reply intelmq/tests/bots/experts/aggregate/testdata/events.txt:1]: `93.184.216.34` has been extracted from the rfc1918 tests.

Comment (wagner-certat) [reply intelmq/tests/bots/experts/aggregate/testdata/events.txt:1]: https://github.com/certtools/intelmq/blob/2190345b7c0699cbcb18d36132d581c306c9ec6f/intelmq/bots/experts/rfc1918/expert.py#L33-L37 doesn't list it. The address is used in the test https://github.com/certtools/intelmq/blob/2190345b7c0699cbcb18d36132d581c306c9ec6f/intelmq/tests/bots/experts/rfc1918/test_expert.py#L16 because it's *not* a reserved address.

File: intelmq/bots/experts/aggregate/expert.py
Comment (wagner-certat) [reply intelmq/bots/experts/aggregate/expert.py:40]: Does this work?
```suggestion
                if datetime.now() <= (datetime.strptime(data['s'], '%Y-%m-%dT%H:%M:%S.%f') + delta):
```

File: intelmq/tests/bots/experts/aggregate/testdata/events.txt
Comment (wagner-certat) [reply intelmq/tests/bots/experts/aggregate/testdata/events.txt:1]: I think this discussion is still open

**Review by wagner-certat — APPROVED** [review]

**Comment by wagner-certat** [comment]
I think it's ready for merge now

**Pull request merged**
